sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1003 @@
+# coding: utf-8
+"""
+Large Margin Nearest Neighbor Classification
+"""
+
+# Author: John Chiotellis <johnyc.code@gmail.com>
+#
+# License: BSD 3 clause
+
+from __future__ import print_function
+from warnings import warn
+
+import numpy as np
+import time
+from scipy.optimize import fmin_l_bfgs_b
+from scipy.sparse import csr_matrix, csc_matrix, spdiags
+
+from ..neighbors import KNeighborsClassifier, NearestNeighbors
+from ..metrics.pairwise import euclidean_distances
+from ..decomposition import PCA
+from ..utils import gen_batches
+from ..utils.extmath import row_norms
+from ..utils.fixes import sp_version, bincount
+from ..utils.random import choice, check_random_state
+from ..utils.multiclass import check_classification_targets
+from ..utils.validation import check_is_fitted, check_array, check_X_y
+from ..exceptions import DataDimensionalityWarning
+
+
+class LargeMarginNearestNeighbor(KNeighborsClassifier):
+    """Distance Metric Learning for Large Margin Classification
+
+    Large margin nearest neighbor classification (LMNN) is a machine learning
+    algorithm for metric learning. It learns a (pseudo-)metric in a
+    supervised fashion to improve the classification accuracy of the k-nearest
+    neighbor rule.
+    The main intuition behind LMNN is to learn a pseudometric under which all
+    data instances in the training set are surrounded by at least k instances
+    that share the same class label. If this is achieved, the leave-one-out
+    error is minimized.
+    This implementation follows closely Kilian Weinberger's MATLAB code found
+    at <https://bitbucket.org/mlcircus/lmnn> which solves the unconstrained
+    problem, finding a linear transformation with L-BFGS instead of solving the
+    constrained problem that finds the globally optimal metric. Different from
+    the paper, the problem solved by this implementation (and the original
+    MATLAB code) is with the squared hinge loss (to make the problem
+    differentiable).
+
+
+    Parameters
+    ----------
+    L : array, shape (n_features_out, n_features_in), optional (default=None)
```

Comment:
We strive to avoid single-letter names, as they are hard to understand out of context. We should strive to find a good name. Maybe "init_metric"?